types/test.d.ts

```diff
@@ -0,0 +1,203 @@
+  isNotEqual: Assertions.NotEqual
+  isNot: Assertions.NotEqual
+  doesNotEqual: Assertions.NotEqual
+  isInequal: Assertions.NotEqual
+
+  same: Assertions.Equal
+  equivalent: Assertions.Equal
+  looseEqual: Assertions.Equal
+  looseEquals: Assertions.Equal
+  deepEqual: Assertions.Equal
+  deepEquals: Assertions.Equal
+  isLoose: Assertions.Equal
+  looseIs: Assertions.Equal
+
+  notSame: Assertions.NotEqual
+  inequivalent: Assertions.NotEqual
+  looseInequal: Assertions.NotEqual
+  notDeep: Assertions.NotEqual
+  deepInequal: Assertions.NotEqual
+  notLoose: Assertions.NotEqual
+  looseNot: Assertions.NotEqual
+
+  strictSame: Assertions.Equal
+  strictEquivalent: Assertions.Equal
+  strictDeepEqual: Assertions.Equal
+  sameStrict: Assertions.Equal
+  deepIs: Assertions.Equal
+  isDeeply: Assertions.Equal
+  isDeep: Assertions.Equal
+  strictDeepEquals: Assertions.Equal
+
+  strictNotSame: Assertions.NotEqual
+  strictInequivalent: Assertions.NotEqual
+  strictDeepInequal: Assertions.NotEqual
+  notSameStrict: Assertions.NotEqual
+  deepNot: Assertions.NotEqual
+  notDeeply: Assertions.NotEqual
+  strictDeepInequals: Assertions.NotEqual
+  notStrictSame: Assertions.NotEqual
+
+  match: Assertions.Match
+  has: Assertions.Match
+  hasFields: Assertions.Match
+  matches: Assertions.Match
+  similar: Assertions.Match
+  like: Assertions.Match
+  isLike: Assertions.Match
+  includes: Assertions.Match
+  include: Assertions.Match
+  contains: Assertions.Match
+
+  notMatch: Assertions.Match
+  dissimilar: Assertions.Match
+  unsimilar: Assertions.Match
+  notSimilar: Assertions.Match
+  unlike: Assertions.Match
+  isUnlike: Assertions.Match
+  notLike: Assertions.Match
+  isNotLike: Assertions.Match
+  doesNotHave: Assertions.Match
+  isNotSimilar: Assertions.Match
+  isDissimilar: Assertions.Match
+
+  type: Assertions.Type
+  isa: Assertions.Type
+  isA: Assertions.Type
+}
+
+export declare namespace Options {
+  export interface Bag {
+    [key: string]: any
+  }
+
+  export interface Pragma {
+    [key: string]: boolean
+  }
+
+  export interface Assert extends Bag {
+    todo?: boolean | string,
+    skip?: boolean | string,
+    diagnostic?: boolean
+  }
+
+  export interface Spawn extends Assert {
+    bail?: boolean,
+    timeout?: number
+  }
+
+  export interface Test extends Assert {
+    timeout?: number,
+    bail?: boolean,
+    autoend?: boolean,
+    buffered?: boolean,
+    jobs?: number,
+    grep?: RegExp[],
+    only?: boolean,
+    runOnly?: boolean
+  }
+}
+
+export declare namespace Assertions {
+  export type Basic = (obj: any, message?: string, extra?: Options.Assert) => boolean
+
+  export type Throws = (fn?: (a: any) => any, expectedError?: Error, message?: string, extra?: Options.Assert) => boolean
+
+  export type DoesNotThrow = (fn?: (a: any) => any, message?: string, extra?: Options.Assert) => boolean
+
+  export type Equal = (found: any, wanted: any, message?: string, extra?: Options.Assert) => boolean
```

Comment:
Using generics will allow us to infer what properties `wanted` should have from `found`.
```suggestion
  export interface Equal {
    <T>(found: T, wanted: T): boolean;
    <T>(found: T, wanted: T, message: string): boolean;
    <T>(found: T, wanted: T, extra: Options.Assert): boolean;
    <T>(found: T, wanted: T, message: string, extra: Options.Assert): boolean;
    <T>(found: T, wanted: T, message?: string, extra?: Options.Assert): boolean;
  }
```

Reply:
Hmm, I'm not sure that use of generics is warranted here @acdibble. What's concerning is that if the actual JS runtime `typeof found` is, say, `"number"` but `typeof wanted` is, say, `"string"` then the generic type `T` would have to be `string | number` to pass static type checking (although guess the TS compiler would likely infer that if not explicitly given).

The purpose of comparing `found` against `wanted` for any kind of equality or inequality assertion is to verify runtime types and values and not compile-time static typings.

Actually, tbh, the first time I had a go at hacking together TypeScript typings for node-tap I dabbled with the generics route as you suggest above but soon abandoned generics for aforementioned reasons. Probably the most accurate TypeScript type for `found` is actually `unknown` as what the test found is always unknown until your test checks it!

The type annotation for `wanted` may as well be `unknown` also as having generic `<T>` arguably adds more technical debt than real value.

Reply:
I see your point. I was using specifically the `deepEqual` method when I was putting together these typings. I think the generic makes sense for that method, as it helps eliminate false assertions or even issues within the code itself.

```typescript
t.deepEqual<number[]>(myFunction(), [0,1,2,3,4]);
```

If `myFunction` doesn't return an array of numbers then you would get a compile-time error which could be helpful. I wasn't considering that this typing is used for several different methods, but I think there would therefore be value to more finely type the different assertion methods.

Reply:
Thanks for engaging @acdibble. I see your point too & would love to see input from others to nail down the best typings in test scenarios!

Reply:
I guess the typings do not necessarily need to reflect what makes sense, but what the untyped code allows. And in this case it would allow to compare apples with oranges. We know the outcome though... So I vote against the generics here. 

Btw. I started to prepare the typings for submission to DefinetlyTyped, including a test. Although I agree with @delvedor, that DT is not the best place for typings, it's still the second best option. Will share my branch here in the upcoming days.